doc/neps/nep-0055-string_dtype.rst

```diff
@@ -0,0 +1,855 @@
+data in the array buffer.
+
+Cython Support and the Buffer Protocol
+++++++++++++++++++++++++++++++++++++++
+
+It's impossible for ``StringDType`` to support the Python buffer protocol, so
+Cython will not support idiomatic typed memoryview syntax for ``StringDType``
+arrays unless special support is added in Cython in the future. We have some
+preliminary ideas for ways to either update the buffer protocol [7]_ or make use
+of the Arrow C data interface [8]_ to expose NumPy arrays for DTypes that don't
+make sense in the buffer protocol, but those efforts will likely not come to
+fruition in time for NumPy 2.0. This means adapting legacy Cython code that uses
+arrays of fixed-width strings to work with ``StringDType`` will be
+non-trivial. Adapting code that worked with object string arrays should be
+straightforward since object arrays aren't supported by the buffer protocol
+either and will likely have no types or have ``object`` type in Cython.
+
+Missing data support
+********************
+
+Missing data can be represented using a sentinel:
+
+  >>> dt = StringDType(na_object=np.nan)
+  >>> arr = np.array(["hello", nan, "world"], dtype=dt)
+  >>> arr
+  array(['hello', nan, 'world'], dtype=StringDType(na_object=nan))
+  >>> arr[1]
+  nan
+  >>> np.isnan(arr[1])
+  True
+  >>> np.isnan(arr)
+  array([False,  True, False])
+  >>> np.empty(3, dtype=dt)
+  array([nan, nan, nan])
+
+We only propose supporting user-provided sentinels. By default, empty arrays
+will be populated with empty strings:
+
+  >>> np.empty(3, dtype=StringDType())
+  array(['', '', ''], dtype=StringDType())
+
+By only supporting user-provided missing data sentinels, we avoid resolving
+exactly how NumPy itself should support missing data and the correct semantics
+of the missing data object, leaving that up to users to decide. However, we *do*
+detect whether the user is providing a NaN-like missing data value, a string
+missing data value, or neither. We explain how we handle these cases below.
+
+A cautious reader may be worried about the complexity of needing to handle three
+different categories of missing data sentinel. The complexity here is reflective
+of the flexibility of object arrays and the downstream usage patterns we've
+found. Some users want comparisons with the sentinel to error, so they use
+``None``. Others want comparisons to succeed and have some kind of meaningful
+ordering, so they use some arbitrary, hopefully unique string. Other users want
+to use something that acts like NaN in comparisons and arithmetic or is
+literally NaN so that NumPy operations that specifically look for exactly NaN
+work and there isn't a need to rewrite missing data handling outside of
+NumPy. We believe it is possible to support all this, but it requires a bit of
+hopefully manageable complexity.
+
+NaN-like Sentinels
+++++++++++++++++++
+
+A NaN-like sentinel returns itself as the result of comparison operations. This
```

Comment:
@ngoldbaum, for the next time you edit the NEP: I think you meant "arithmetic operations" here, not "comparison operations".  *Comparing* a floating point nan with something else always returns False, not itself.